packages/injected/src/clock.ts

```diff
@@ -192,38 +202,63 @@ export class ClockController {
   }
 
   private _updateRealTimeTimer() {
-    if (!this._realTime) {
-      this._currentRealTimeTimer?.dispose();
-      this._currentRealTimeTimer = undefined;
+    if (this._currentRealTimeTimer?.promise) {
+      // In progress, safe to return as it will call itself once promise is resolved.
       return;
     }
 
     const firstTimer = this._firstTimer();
 
     // Either run the next timer or move time in 100ms chunks.
     const callAt = Math.min(firstTimer ? firstTimer.callAt : this._now.ticks + maxTimeout, this._now.ticks + 100) as Ticks;
-    if (this._currentRealTimeTimer && this._currentRealTimeTimer.callAt < callAt)
-      return;
 
-    if (this._currentRealTimeTimer) {
-      this._currentRealTimeTimer.dispose();
+    if (this._currentRealTimeTimer && this._currentRealTimeTimer.callAt > callAt) {
+      // Cancel and reschedule to sooner time.
```

Comment:
I think we must `return` otherwise.

Currently, if there is a timer scheduled for some earlier time, we'll schedule a second one right away and they might run concurrently.